### ISSUE TITLE
Libtorch/c++ GPU inferance

### ISSUE THREAD

**Issue by CURTLab** [ultralytics/ultralytics#7744]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Integrations

### Bug

Pytorch version: 2.1.2
Layer Device: cuda:0
Input Device: cuda:0

Error: The following operation failed in the TorchScript interpreter.
Traceback of TorchScript, serialized code (most recent call last):
  File "code/__torch__/ultralytics/nn/tasks/___torch_mangle_1331.py", line 74, in forward
    _35 = (_18).forward(act, _34, )
    _36 = (_20).forward((_19).forward(act, _35, ), _29, )
    _37 = (_22).forward(_33, _35, (_21).forward(act, _36, ), )
           ~~~~~~~~~~~~ <--- HERE
    return _37
  File "code/__torch__/ultralytics/nn/modules/head/___torch_mangle_1329.py", line 46, in forward
    anchor_points = torch.unsqueeze(CONSTANTS.c0, 0)
    lt, rb, = torch.chunk(_14, 2, 1)
    x1y1 = torch.sub(anchor_points, lt)
           ~~~~~~~~~ <--- HERE
    x2y2 = torch.add(anchor_points, rb)
    c_xy = torch.div(torch.add(x1y1, x2y2), CONSTANTS.c1)

Traceback of TorchScript, original code (most recent call last):
c:\Program Files\Python310\lib\site-packages\ultralytics\utils\tal.py(266): dist2bbox
c:\Program Files\Python310\lib\site-packages\ultralytics\nn\modules\head.py(53): forward
c:\Program Files\Python310\lib\site-packages\torch\nn\modules\module.py(1176): _slow_forward
c:\Program Files\Python310\lib\site-packages\torch\nn\modules\module.py(1192): _call_impl
c:\Program Files\Python310\lib\site-packages\ultralytics\nn\tasks.py(76): _predict_once
c:\Program Files\Python310\lib\site-packages\ultralytics\nn\tasks.py(58): predict
c:\Program Files\Python310\lib\site-packages\ultralytics\nn\tasks.py(41): forward
c:\Program Files\Python310\lib\site-packages\torch\nn\modules\module.py(1176): _slow_forward
c:\Program Files\Python310\lib\site-packages\torch\nn\modules\module.py(1192): _call_impl
c:\Program Files\Python310\lib\site-packages\torch\jit\_trace.py(957): trace_module
c:\Program Files\Python310\lib\site-packages\torch\jit\_trace.py(753): trace
c:\Program Files\Python310\lib\site-packages\ultralytics\engine\exporter.py(298): export_torchscript
c:\Program Files\Python310\lib\site-packages\ultralytics\engine\exporter.py(116): outer_func
c:\Program Files\Python310\lib\site-packages\ultralytics\engine\exporter.py(202): __call__
c:\Program Files\Python310\lib\site-packages\torch\autograd\grad_mode.py(26): decorate_context
c:\Program Files\Python310\lib\site-packages\ultralytics\engine\model.py(309): export
C:\Users\ownha\AppData\Local\Temp\ipykernel_25752\1005294438.py(1): <module>
c:\Program Files\Python310\lib\site-packages\IPython\core\interactiveshell.py(3430): run_code
c:\Program Files\Python310\lib\site-packages\IPython\core\interactiveshell.py(3341): run_ast_nodes
c:\Program Files\Python310\lib\site-packages\IPython\core\interactiveshell.py(3168): run_cell_async
c:\Program Files\Python310\lib\site-packages\IPython\core\async_helpers.py(129): _pseudo_sync_runner
c:\Program Files\Python310\lib\site-packages\IPython\core\interactiveshell.py(2970): _run_cell
c:\Program Files\Python310\lib\site-packages\IPython\core\interactiveshell.py(2941): run_cell
c:\Program Files\Python310\lib\site-packages\ipykernel\zmqshell.py(532): run_cell
c:\Program Files\Python310\lib\site-packages\ipykernel\ipkernel.py(378): do_execute
c:\Program Files\Python310\lib\site-packages\ipykernel\kernelbase.py(700): execute_request
c:\Program Files\Python310\lib\site-packages\ipykernel\kernelbase.py(383): dispatch_shell
c:\Program Files\Python310\lib\site-packages\ipykernel\kernelbase.py(496): process_one
c:\Program Files\Python310\lib\site-packages\ipykernel\kernelbase.py(510): dispatch_queue
c:\Program Files\Python310\lib\asyncio\events.py(80): _run
c:\Program Files\Python310\lib\asyncio\base_events.py(1863): _run_once
c:\Program Files\Python310\lib\asyncio\base_events.py(597): run_forever
c:\Program Files\Python310\lib\site-packages\tornado\platform\asyncio.py(212): start
c:\Program Files\Python310\lib\site-packages\ipykernel\kernelapp.py(704): start
c:\Program Files\Python310\lib\site-packages\traitlets\config\application.py(983): launch_instance
c:\Program Files\Python310\lib\site-packages\ipykernel_launcher.py(12): <module>
c:\Program Files\Python310\lib\runpy.py(75): _run_code
c:\Program Files\Python310\lib\runpy.py(191): _run_module_as_main
RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu!

### Environment

Ultralytics YOLOv8.0.227 🚀 Python-3.10.11 torch-2.0.1+cu117 CUDA:0 (NVIDIA GeForce MX130, 2048MiB)
Setup complete ✅ (8 CPUs, 31.9 GB RAM, 1238.1/1863.0 GB disk)

### Minimal Reproducible Example

```c++
std::cout << "Pytorch version: " << TORCH_VERSION_MAJOR << "." << TORCH_VERSION_MINOR << "." << TORCH_VERSION_PATCH << std::endl;

try {
	torch::InferenceMode guard(true);

	d->model = torch::jit::load(modelPath.toStdString());
	d->model.eval();
	d->model.to(d->device);

	d->options = F::InterpolateFuncOptions().size(std::vector<int64_t>{modelSize.height(), modelSize.width()}).mode(torch::kNearest);

	auto layer = *d->model.parameters().begin();
	// get data type
	d->dtype = layer.data().dtype().toScalarType();
	if (layer.data().device().type() != d->device.type()) {
		d->lastErrorMessage = "Invalid device type!";
		std::cout << __FUNCTION__ ": Error loading model: " << qPrintable(d->lastErrorMessage) << std::endl;
		return -1;
	}
	std::cout << "Layer Device: " << layer.data().device() << std::endl;

	// warmup
	auto input = torch::zeros({1, 3, modelSize.height(), modelSize.width()}, torch::TensorOptions(d->dtype).device(d->device));
	std::cout << "Input Device: " << input.device() << std::endl;
	d->model.forward({input}).toTensor();
} catch (const c10::Error& e) {
	d->lastErrorMessage = QString("Torch error: %1").arg(e.what());
	std::cout << __FUNCTION__ ": Error loading model: " << qPrintable(d->lastErrorMessage) << std::endl;
	return -1;
} catch (const std::exception& e) {
	d->lastErrorMessage = QString("Error: %1").arg(e.what());
	std::cout << __FUNCTION__ ": Error loading model: " << qPrintable(d->lastErrorMessage) << std::endl;
	return -1;
}
```

### Additional

I have trained a yolov8 network for custom data and wanted to run this network using libtorch c++. CPU works find but GPU always crashes even during warmup. 

I exported the model using: ```path = model.export(format="torchscript")```

I also tried to jit.trace the model before exporting. Everything works fine in python but as soon as I use libtorch/c++ with GPU I get the above mentioned error.

model.task_map: ```{'classify': {'model': <class 'ultralytics.nn.tasks.ClassificationModel'>, 'trainer': <class 'ultralytics.models.yolo.classify.train.ClassificationTrainer'>, 'validator': <class 'ultralytics.models.yolo.classify.val.ClassificationValidator'>, 'predictor': <class 'ultralytics.models.yolo.classify.predict.ClassificationPredictor'>}, 'detect': {'model': <class 'ultralytics.nn.tasks.DetectionModel'>, 'trainer': <class 'ultralytics.models.yolo.detect.train.DetectionTrainer'>, 'validator': <class 'ultralytics.models.yolo.detect.val.DetectionValidator'>, 'predictor': <class 'ultralytics.models.yolo.detect.predict.DetectionPredictor'>}, 'segment': {'model': <class 'ultralytics.nn.tasks.SegmentationModel'>, 'trainer': <class 'ultralytics.models.yolo.segment.train.SegmentationTrainer'>, 'validator': <class 'ultralytics.models.yolo.segment.val.SegmentationValidator'>, 'predictor': <class 'ultralytics.models.yolo.segment.predict.SegmentationPredictor'>}, 'pose': {'model': <class 'ultralytics.nn.tasks.PoseModel'>, 'trainer': <class 'ultralytics.models.yolo.pose.train.PoseTrainer'>, 'validator': <class 'ultralytics.models.yolo.pose.val.PoseValidator'>, 'predictor': <class 'ultralytics.models.yolo.pose.predict.PosePredictor'>}}```
model.model.args['imgsz']: ```1376```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by CURTLab** [comment]
I have found a similar issue for yolo-v5 [RuntimeError: The following operation failed in the TorchScript interpreter](https://github.com/ultralytics/yolov5/issues/5070). Here the described similar error messages for python and to check if the input shape/type and model device is correct. I did the same and check all model layers using `d->model.parameters()`. All layers were on 'cuda:0' (see also terminal output Layer Device & Input Device). Are there some hidden things in the yolo-v8 model that are not moved to the GPU using `model.to(d->device)`?

**Comment by glenn-jocher** [comment]
Hello @CURTLab! The error you're encountering suggests that there might be certain elements within the model that are not being correctly moved to the GPU. While `model.to(d->device)` should ideally move all parameters and buffers to the specified device, there could be other tensors created during the forward pass that are not automatically placed on the GPU.

In YOLOv8, as with other complex models, there may be tensors created within the forward function or within certain layers that need to be explicitly moved to the correct device. This can include intermediate tensors, constants, or tensors used for post-processing steps.

To resolve this, ensure that any tensor created during the model's forward pass or within any custom layers is also moved to the GPU. This includes tensors that might be part of the model's architecture but are not registered as parameters or buffers.

If you continue to face issues, please provide a detailed report on the Ultralytics YOLOv8 GitHub repository, and we'll investigate further. Remember to include all relevant information, including the full error message, the environment setup, and the steps you've taken to debug the issue. We're here to help! 🛠️

**Comment by glenn-jocher** [comment]
Hi @CURTLab, that's fantastic news! 🎉 I'm glad to hear you found a solution. Indeed, loading the model directly onto the desired device can sometimes be more reliable than moving it afterward with `model.to(d->device)`. 

It might be beneficial for the PyTorch community if you share your findings in an issue on their repository. It could help improve the documentation or even lead to enhancements in future releases. Thanks for contributing to the community, and happy coding with YOLOv8 and LibTorch/C++! 😊👍

**Comment by UltralyticsAssistant** [comment]
A potential fix for this issue has just been merged in [PR #16390](https://github.com/ultralytics/ultralytics/pull/16390): **Fix CUDA Error in YOLO Libtorch C++ Implementation**. The main update involves specifying the device during model loading in the YOLOv8 C++ inference example using `torch::jit::load(model_path, device)`. This change helps ensure the model is loaded directly onto the correct device (CPU or GPU), which should reduce device mismatch errors and improve compatibility and performance—especially for GPU users.

Big thanks to @ssapsu for authoring the PR, with valuable contributions from @Myyura, @Burhan-Q, @shikhindahikar, @glenn-jocher, @ambitious-octopus, and @yang0331.

To test if this resolves your issue, you can install the latest code directly with:
```
pip install git+https://github.com/ultralytics/ultralytics.git@main
```
Alternatively, you may wait for the next official PyPI release, where this fix will be included.

Please let us know if this update resolves your issue! Thank you for bringing this to our attention, and feel free to share any further feedback if the problem persists. 🙏
